### PR TITLE
Separate system-probe config from datadog config

### PR DESCRIPTION
**Pull request by brycekahle**
### What does this PR do?

Separates the system-probe configuration into a distinct global named `config.SystemProbe`.

### Motivation

Because the `system-probe.yaml` config controls a process running as root, it should be extra clear where configuration is coming from. Previously, the configuration from both `datadog.yaml` and `system-probe.yaml` was being flattened into the same namespace, which caused a lot of confusion.

This also prevented accurate detection of a user using incorrect configuration. For example, trying to use `system_probe_config:` in `datadog.yaml`, is incorrect, but would not be warned about in a log message.

### Additional Notes

Some of the configuration overriding logic was applying even to the system-probe config, so I gated a lot of it using a new function `config.IsKnown`. This should prevent the overrides from applying to configurations which do not have those keys.

I also added a `sysprobeconfig` component for use in agent components.

I removed some extraneous log messages from the system-probe `config.New` function, because those were being output even in the `agent`. I moved equivalent messages to be output in `system-probe` upon startup.

### Describe how to test/QA your changes

I updated all the tests, and did some manual checking of subcommands like `agent status`, `agent check`, `system-probe config`, `system-probe debug`, etc.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: cmd/agent/subcommands/run/dependent_services_windows.go [hunk 1]
```diff
@@ -22,43 +23,47 @@ type serviceInitFunc func() (err error)
...
 		name: "process",
-		configKeys: []string{
-			"process_config.enabled",
-			"process_config.process_collection.enabled",
-			"process_config.container_collection.enabled",
-			"process_config.process_discovery.enabled",
-			"network_config.enabled",
-			"system_probe_config.enabled",
+		configKeys: map[config.Config]string{
+			config.Datadog:     "process_config.enabled",
```
Comment (leeavital) [inline cmd/agent/subcommands/run/dependent_services_windows.go:44]: I might be missing some go lore, but won't this map only wind up with two entries ? you might want to reverse the keys and values

https://play.golang.com/p/zeaxiOxX9kQ

File: cmd/system-probe/app/debug.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ func init() {
 
 var (
 	debugCommand = &cobra.Command{
-		Use:   "debug [state]",
+		Use:   "debug [module] [state]",
```
Comment (leeavital) [inline cmd/system-probe/app/debug.go:28]: How is this related to config?

File: comp/core/sysprobeconfig/component.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	// Set sets the given config value
+	Set(key string, value interface{})
+}
+
+var componentName = computil.GetComponentName()
+
+// Module defines the fx options for this component.
+var Module = fx.Module(
+	componentName,
+	fx.Provide(newConfig),
```
Comment (leeavital) [inline comp/core/sysprobeconfig/component.go:127]: This function doesn't seem to call InitSystemProbeConfig -- is that okay ?

File: cmd/agent/subcommands/run/dependent_services_windows.go (hunk 1 above)
Comment (brycekahle) [reply cmd/agent/subcommands/run/dependent_services_windows.go:44]: Derp, you are right.

File: comp/core/sysprobeconfig/component.go (hunk 3 above)
Comment (brycekahle) [reply comp/core/sysprobeconfig/component.go:127]: Yes, because for now it is still using the `config.SystemProbe` global.

File: cmd/system-probe/app/debug.go (hunk 2 above)
Comment (brycekahle) [reply cmd/system-probe/app/debug.go:28]: It was actually broken before and I discovered it during testing of the config. It wasn't allowing access to any of the module-specific debug endpoints.

File: cmd/system-probe/config/config.go [hunk 4]
```diff
@@ -72,58 +73,40 @@ type Config struct {
...
+	// If `!failOnMissingFile`, do not issue an error if we cannot find the default config file.
 	var e viper.ConfigFileNotFoundError
-	if err != nil {
-		if errors.As(err, &e) || errors.Is(err, os.ErrNotExist) {
-			log.Infof("no config exists at %s, ignoring...", configPath)
+	if err != nil && (!errors.As(err, &e) || configPath != "") {
+		// special-case permission-denied with a clearer error message
+		if errors.Is(err, fs.ErrPermission) {
+			if runtime.GOOS == "windows" {
+				err = fmt.Errorf(`cannot access the system-probe config file (%w); try running the command in an Administrator shell"`, err)
```
Comment (xornivore) [inline cmd/system-probe/config/config.go:96]: Do we still consider error wraps with `%w` useful?

File: cmd/system-probe/config/config.go [hunk 5]
```diff
@@ -182,42 +165,36 @@ func load(configPath string) (*Config, error) {
...
-		log.Info("network_config not found, but system-probe was enabled, enabling network module by default")
+		log.Info("`system_probe_config.enabled` is deprecated, enable NPM with `network_config.enabled` instead")
 		c.EnabledModules[NetworkTracerModule] = struct{}{}
 		// ensure others can key off of this single config value for NPM status
 		cfg.Set("network_config.enabled", true)
 		npmEnabled = true
 	}
 
 	if !npmEnabled && usmEnabled {
-		log.Info("service_monitoring.enabled detected: enabling system-probe with network module running.")
```
Comment (xornivore) [inline cmd/system-probe/config/config.go:203]: perhaps you may want to keep this in one release?

File: cmd/system-probe/modules/network_tracer.go [hunk 6]
```diff
@@ -50,7 +49,12 @@ var NetworkTracer = module.Factory{
 			return nil, fmt.Errorf("%w: %s", ErrSysprobeUnsupported, msg)
 		}
 
-		log.Infof("Creating tracer for: %s", filepath.Base(os.Args[0]))
+		if ncfg.NPMEnabled {
+			log.Info("network performance monitoring (NPM) enabled")
```
Comment (xornivore) [inline cmd/system-probe/modules/network_tracer.go:53]:
```suggestion
			log.Info("enabling network performance monitoring (NPM)")
```

File: cmd/system-probe/modules/network_tracer.go [hunk 7]
```diff
@@ -50,7 +49,12 @@ var NetworkTracer = module.Factory{
...
+		}
+		if ncfg.ServiceMonitoringEnabled {
+			log.Info("universal service monitoring (USM) enabled")
```
Comment (xornivore) [inline cmd/system-probe/modules/network_tracer.go:56]:
```suggestion
			log.Info("enabling universal service monitoring (USM)")
```

File: comp/core/sysprobeconfig/component.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	"time"
+
+	"go.uber.org/fx"
+
+	"github.com/DataDog/datadog-agent/comp/internal/computil"
+	"github.com/DataDog/datadog-agent/pkg/config"
+)
+
+// Component is the component type.
+type Component interface {
```
Comment (xornivore) [inline comp/core/sysprobeconfig/component.go:28]: Any way to reuse a common interface here?

File: pkg/process/config/config_test.go [hunk 9]
```diff
@@ -115,29 +115,41 @@ func TestBlacklist(t *testing.T) {
...
 	// Reset the config
 	config.Datadog = config.NewConfig("datadog", "DD", strings.NewReplacer(".", "_"))
+	config.InitConfig(config.Datadog)
 
-	f, err := ioutil.TempFile("", "yamlConfigTest*.yaml")
-	defer os.Remove(f.Name())
-	assert.NoError(t, err)
+	f, err := os.CreateTemp(t.TempDir(), "yamlConfigTest*.yaml")
+	require.NoError(t, err)
+	defer f.Close()
```
Comment (xornivore) [inline pkg/process/config/config_test.go:152]: Mind clarifying this change?

File: pkg/process/config/config.go [hunk 10]
```diff
@@ -281,15 +280,6 @@ func loadEnvVariables() {
...
 	}
-
-	if v := os.Getenv("DD_ORCHESTRATOR_ADDITIONAL_ENDPOINTS"); v != "" {
-		endpoints := make(map[string][]string)
-		if err := json.Unmarshal([]byte(v), &endpoints); err != nil {
-			log.Errorf(`Could not parse DD_ORCHESTRATOR_ADDITIONAL_ENDPOINTS: %v. It must be of the form '{"https://process.agent.datadoghq.com": ["apikey1", ...], ...}'.`, err)
-		} else {
-			config.Datadog.Set("orchestrator_explorer.orchestrator_additional_endpoints", endpoints)
-		}
-	}
```
Comment (xornivore) [inline pkg/process/config/config.go:292]: You may want to call this change out in the testing recommendation.

File: pkg/config/environment_detection.go [hunk 11]
```diff
@@ -114,7 +114,9 @@ func DetectFeatures() {
 			}
 		}
 
-		log.Infof("Features detected from environment: %v", newFeatures)
+		if len(newFeatures) > 0 {
+			log.Infof("Features detected from environment: %v", newFeatures)
```
Comment (clamoriniere) [inline pkg/config/environment_detection.go:118]: IMO, even if the `newFeature` slice is empty it can be good to keep this log line.

maybe
```suggestion
			log.Infof("%d Features detected from environment: %v", len(newFeatures), newFeatures)
```

**Review by KSerrania — APPROVED** [review]
LGTM for Agent Platform-owned files.

**Review by mfpierre — APPROVED** [review]
LGTM for container-app

File: pkg/process/config/config.go [hunk 12]
```diff
@@ -281,15 +280,6 @@ func loadEnvVariables() {
 			config.Datadog.Set(variable.cfg, v)
 		}
 	}
-
-	if v := os.Getenv("DD_ORCHESTRATOR_ADDITIONAL_ENDPOINTS"); v != "" {
```
Comment (just-chillin) [inline pkg/process/config/config.go:285]: What is this being replaced by? I think we still need to support this env var

**Review by just-chillin — APPROVED** [review]
LGTM - @xornivore brought up some really good points, once those are addressed it's good to merge as far as I'm concerned

File: pkg/process/config/config.go (hunk 12 above)
Comment (mfpierre) [reply pkg/process/config/config.go:285]: Oh wait, yeah thanks for calling this out @just-chillin it was not part of our owned files 😬 indeed this is still actively used (both in the cluster-agent and process-agent)

File: pkg/config/config.go [hunk 13]
```diff
@@ -1017,6 +1020,8 @@ func InitConfig(config Config) {
 	config.BindEnv("orchestrator_explorer.max_message_bytes")
 	config.BindEnv("orchestrator_explorer.orchestrator_dd_url")
 	config.BindEnv("orchestrator_explorer.orchestrator_additional_endpoints")
+	// add additional env var name
+	config.BindEnv("orchestrator_explorer.orchestrator_additional_endpoints", "DD_ORCHESTRATOR_ADDITIONAL_ENDPOINTS")
```
Comment (mfpierre) [inline pkg/config/config.go:1024]: This is a duplicate from the above line

File: pkg/process/config/config.go (hunk 12 above)
Comment (mfpierre) [reply pkg/process/config/config.go:285]: Why would you want to remove this bit?

**Review by mfpierre — CHANGES_REQUESTED** [review]

File: cmd/agent/subcommands/run/dependent_services_windows.go [hunk 14]
```diff
@@ -93,7 +98,13 @@ func (s *Servicedef) Start() error {
...
 	defer m.Disconnect()
 
-	hSvc, err := windows.OpenService(m.Handle, syscall.StringToUTF16Ptr(s.serviceName),
+	snptr, err := syscall.UTF16PtrFromString(s.serviceName)
+	if err != nil {
+		log.Warnf("Failed to get service name %v", err)
+		return fmt.Errorf("could not create service name pointer: %s", err)
+	}
+
+	hSvc, err := windows.OpenService(m.Handle, snptr,
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply cmd/agent/subcommands/run/dependent_services_windows.go:107]: The previous function was marked deprecated, so I fixed it here. I can pull this out if wanted.

File: cmd/system-probe/config/config.go (hunk 4 above)
Comment (brycekahle) [reply cmd/system-probe/config/config.go:96]: They are only useful is something higher in the call stack needs to check for a specific error. It is a reasonable default though.

File: cmd/system-probe/config/config.go (hunk 5 above)
Comment (brycekahle) [reply cmd/system-probe/config/config.go:203]: These logs got moved to system-probe at runtime, so they don't get logged for other processes that consume the system-probe config.

File: pkg/process/config/config_test.go (hunk 9 above)
Comment (brycekahle) [reply pkg/process/config/config_test.go:152]: `ioutil.TempFile` is considered deprecated as of 1.17. It is calling `os.CreateTemp` underneath. Using `t.TempDir()` ensures that the temp file is automatically cleaned up when the test completes.

File: pkg/config/environment_detection.go (hunk 11 above)
Comment (brycekahle) [reply pkg/config/environment_detection.go:118]: 👍

File: pkg/config/config.go (hunk 13 above)
Comment (brycekahle) [reply pkg/config/config.go:1024]: Intentional duplicate to also bind the `DD_ORCHESTRATOR_ADDITIONAL_ENDPOINTS` name in addition to the default.

File: pkg/process/config/config.go (hunk 12 above)
Comment (brycekahle) [reply pkg/process/config/config.go:285]: The custom unmarshaling isn't needed. `config.Datadog.GetStringMapStringSlice` will handle that format.

File: cmd/security-agent/api/agent/agent.go [hunk 15]
```diff
@@ -49,7 +49,7 @@ func (a *Agent) SetupHandlers(r *mux.Router) {
 	r.HandleFunc("/stop", a.stopAgent).Methods("POST")
 	r.HandleFunc("/status", a.getStatus).Methods("GET")
 	r.HandleFunc("/status/health", a.getHealth).Methods("GET")
-	r.HandleFunc("/config", settingshttp.Server.GetFull("")).Methods("GET")
+	r.HandleFunc("/config", settingshttp.Server.GetFull(config.Datadog, "")).Methods("GET")
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply cmd/security-agent/api/agent/agent.go]: Updated.

File: cmd/process-agent/app/check.go [hunk 16]
```diff
@@ -82,7 +78,7 @@ func runCheckCmd(cmd *cobra.Command, args []string) error {
 	}
 
 	// For system probe, there is an additional config file that is shared with the system-probe
-	syscfg, err := sysconfig.Merge(sysprobePath)
+	syscfg, err := sysconfig.New(sysprobePath)
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply cmd/process-agent/app/check.go:81]: 👍

File: comp/core/sysprobeconfig/component.go (hunk 8 above)
Comment (brycekahle) [reply comp/core/sysprobeconfig/component.go:28]: I could, but my understanding was that it should be separate due to how `fx` works. Is that right @djmitche ?

File: pkg/process/config/config.go (hunk 10 above)
Comment (brycekahle) [reply pkg/process/config/config.go:292]: There is an automated test for this env var already https://github.com/DataDog/datadog-agent/blob/2733839ba79090a9f0fa2d038b1ba56190527948/pkg/process/config/config_test.go#L376

File: pkg/process/config/config.go (hunk 12 above)
Comment (brycekahle) [reply pkg/process/config/config.go:285]: This config is still read here: https://github.com/DataDog/datadog-agent/blob/1d75034f1cad5d3a853183567796feea97aa68ce/pkg/orchestrator/config/config.go#L122

**Review by djmitche — APPROVED** [review]

File: pkg/process/config/config.go (hunk 12 above)
Comment (mfpierre) [reply pkg/process/config/config.go:285]: make sense 👍

File: pkg/config/config.go (hunk 13 above)
Comment (mfpierre) [reply pkg/config/config.go:1024]: indeed 👍

**Review by mfpierre — APPROVED** [review]

File: cmd/agent/subcommands/run/command.go [hunk 17]
```diff
@@ -143,7 +144,7 @@ func Commands(globalParams *command.GlobalParams) []*cobra.Command {
 // run starts the main loop.
 //
 // This is exported because it also used from the deprecated `agent start` command.
-func run(log log.Component, config config.Component, cliParams *cliParams) error {
+func run(log log.Component, config config.Component, sysconfig sysprobeconfig.Component, cliParams *cliParams) error {
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply cmd/agent/subcommands/run/command.go:147]: That would ensure the parameter `sysprobeconfig` shadows the entire package `sysprobeconfig` within the function. Is that desired?

File: cmd/agent/subcommands/run/dependent_services_windows.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (xornivore) [reply cmd/agent/subcommands/run/dependent_services_windows.go:107]: No this is fine, just trying to map out what's related 😄

File: pkg/process/config/config_test.go (hunk 9 above)
Comment (xornivore) [reply pkg/process/config/config_test.go:152]: Ah, nice, thank you.

**Review by xornivore — APPROVED** [review]
Nice, thank you for clarifying all the bits 🙇

File: comp/core/sysprobeconfig/setup.go [hunk 18]
```diff
@@ -3,17 +3,14 @@
...
 // Copyright 2016-present Datadog, Inc.
 
-//go:build process
-// +build process
-
-package common
+package sysprobeconfig
 
 import (
 	sysconfig "github.com/DataDog/datadog-agent/cmd/system-probe/config"
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply comp/core/sysprobeconfig/setup.go:9]: I started to do this, but it churns a lot of code. I can address this in a future PR.

File: cmd/agent/subcommands/run/command.go [hunk 19]
```diff
@@ -423,7 +424,7 @@ func startAgent(cliParams *cliParams) error {
 		}
 	}
 
-	if err = common.SetupSystemProbeConfig(cliParams.GlobalParams.SysProbeConfFilePath); err != nil {
+	if _, err := sysconfig.New(cliParams.SysProbeConfFilePath); err != nil {
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply cmd/agent/subcommands/run/command.go:427]: I decided to just remove this. The status component uses whether or not the system-probe is *enabled* to include it, not whether or not the config is loaded.

**Review by usamasaqib — APPROVED** [review]

**Review by safchain — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

File: pkg/config/environment_detection.go [hunk 20]
```diff
@@ -114,7 +114,9 @@ func DetectFeatures() {
 			}
 		}
 
-		log.Infof("Features detected from environment: %v", newFeatures)
+		if len(newFeatures) > 0 {
```
Comment (clamoriniere) [inline pkg/config/environment_detection.go:117]: I think we should remove the `if` to also get the log line when any feature is enable

File: pkg/config/environment_detection.go [hunk 21]
```diff
@@ -114,7 +114,9 @@ func DetectFeatures() {
...
+			log.Infof("%d Features detected from environment: %v", len(newFeatures), newFeatures)
+		}
```
Comment (clamoriniere) [inline pkg/config/environment_detection.go:119]: I think we should remove the `if` to also get the log line when any feature is enable

```suggestion
		log.Infof("%d Features detected from environment: %v", len(newFeatures), newFeatures)
```

**Review by leeavital — APPROVED** [review]

**Review by clamoriniere — APPROVED** [review]
